modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/db/wal/WalRolloverOnStopTest.java

```diff
@@ -0,0 +1,168 @@
+import org.junit.Test;
+
+import static org.apache.ignite.testframework.GridTestUtils.runAsync;
+
+/**
+ * This tests check that the following scenario will works correctly.
+ */
+public class WalRolloverOnStopTest extends GridCommonAbstractTest {
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        return super.getConfiguration(igniteInstanceName)
+            .setDataStorageConfiguration(new DataStorageConfiguration()
+            .setWalAutoArchiveAfterInactivity(1500L)
+            .setDefaultDataRegionConfiguration(new DataRegionConfiguration()
+                .setPersistenceEnabled(true)));
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTestsStarted() throws Exception {
+        cleanPersistenceDir();
+    }
+
+    /**
+     * Test scenario:
+     *
+     * 0. {@link DataStorageConfiguration#getWalAutoArchiveAfterInactivity()} > 0.
+     * 1. Node is gracefully stopping using {@link G#stop(String, boolean)}.
+     * 2. T0: {@code Checkpointer#doCheckpoint()} execute last checkpoint on stop and freeze.
+     * 3. T1: Rollover segment after inactivity timeout.
+     * 4. T2: Archive segment.
+     *
+     * After restart WAL should log in the next segment.
+     * */
+    @Test
+    public void testWallRollover() throws Exception {
+        for (int i = 0; i < 2; i++) {
+            IgniteEx ign = startGrid(0);
+
+            GridCacheDatabaseSharedManager db =
+                (GridCacheDatabaseSharedManager)ign.context().cache().context().database();
+
+            SegmentAware aware = GridTestUtils.getFieldValue(ign.context().cache().context().wal(), "segmentAware");
+
+            ign.cluster().state(ClusterState.ACTIVE);
+
+            IgniteCache<Integer, Integer> cache = ign.getOrCreateCache("my-cache");
+
+            CountDownLatch waitAfterCp = new CountDownLatch(1);
+            AtomicLong cntr = new AtomicLong(0);
+
+            db.addCheckpointListener(new WaitOnLastCheckpoint(ign, waitAfterCp, cntr, getTestTimeout()));
+
+            for (int j = i * 3; j < (i + 1) * 3; j++)
+                cache.put(j, j);
+
+            long cutIdx = aware.curAbsWalIdx();
+
+            runAsync(() -> {
+                try {
+                    aware.awaitSegmentArchived(cutIdx);
+
+                    cntr.incrementAndGet();
+                }
+                catch (IgniteInterruptedCheckedException e) {
+                    throw new IgniteException(e);
+                }
+                finally {
+                    waitAfterCp.countDown();
+                }
+            });
+
+            G.stop(ign.name(), false);
+
+            // Checkpoint will happens two time because of segment archivation.
+            assertEquals("Should successfully wait for current segment archivation", 3, cntr.get());
+        }
+    }
+
+    /** */
+    private static class WaitOnLastCheckpoint implements CheckpointListener {
```

Comment:
I think the anonymous class can be used as a test listener. This will simplify the code itself.